File: compare_exactly_two_steps.m
% Force exactly TWO time steps for comparison
clear all;
close all;

setup_paths;

% Parameters
Np = 20;
Kn = 1.0;
Ma = 0.0;
flag2D = 0;
CFL = 0.5;
dx = 2.0 / Np;
dy = 2.0 / Np;

% Create params struct - set tmax to stop after ~2 steps
params = struct();
params.Np = Np;
params.tmax = 0.04;  % Should give us ~2 steps
params.Kn = Kn;
params.Ma = Ma;
params.flag2D = flag2D;
params.CFL = CFL;
params.dx = dx;
params.dy = dy;
params.Nmom = 35;
params.nnmax = 1000;
params.dtmax = 0.025;
params.rhol = 1.0;
params.rhor = 0.01;
params.T = 1.0;
params.r110 = 0.0;
params.r101 = 0.0;
params.r011 = 0.0;
params.symmetry_check_interval = 0;
params.enable_memory_tracking = false;

fprintf('Running MATLAB simulation for exactly TWO steps...\n');
script_dir = pwd;
[M_final, final_time, time_steps, grid_out] = simulation_runner(params, script_dir);

% Package results
result = struct();
result.final_moments = M_final;
result.final_time = final_time;
result.time_steps = time_steps;

% Save results
save('matlab_exactly_two_steps.mat', 'result', '-v7.3');

fprintf('\n=== MATLAB Results ===\n');
fprintf('Steps completed: %d\n', time_steps);
fprintf('Final time: %.15e\n', final_time);

fprintf('\nFinal M(10,10,1:5):\n');
for i = 1:5
    fprintf('  M(10,10,%d) = %.15e\n', i, M_final(10,10,i));
end

fprintf('\nResults saved\n');

File: RodneyHQMOM.jl/compare_exactly_two_steps.jl
#!/usr/bin/env julia
# Force exactly TWO time steps for comparison

using RodneyHQMOM
using MAT
using Printf

println("Running Julia simulation for exactly TWO steps...")
result = run_simulation(Np=20, tmax=0.04, verbose=true)

M_final = result[:M]

println("\n=== Julia Results ===")
println("Steps completed: ", result[:time_steps])
@printf("Final time: %.15e\n", result[:final_time])

println("\nFinal M[10,10,1:5]:")
for i = 1:5
    @printf("  M[10,10,%d] = %.15e\n", i, M_final[10,10,i])
end

# Save Julia results
matwrite("julia_exactly_two_steps.mat", Dict(
    "final_moments" => M_final,
    "final_time" => result[:final_time],
    "time_steps" => result[:time_steps]
))

# Compare with MATLAB
if isfile("../matlab_exactly_two_steps.mat")
    println("\n" * "="^60)
    println("COMPARISON")
    println("="^60)
    
    matlab_data = matread("../matlab_exactly_two_steps.mat")
    matlab_M = matlab_data["result"]["final_moments"]
    matlab_t = matlab_data["result"]["final_time"]
    matlab_steps = Int(matlab_data["result"]["time_steps"])
    
    julia_M = M_final
    julia_t = result[:final_time]
    julia_steps = result[:time_steps]
    
    @printf("\nSteps: MATLAB=%d, Julia=%d\n", matlab_steps, julia_steps)
    @printf("Final time: MATLAB=%.15e, Julia=%.15e\n", matlab_t, julia_t)
    
    println("\nPoint (10,10) comparison:")
    for i = 1:5
        m_val = matlab_M[10,10,i]
        j_val = julia_M[10,10,i]
        diff = abs(m_val - j_val)
        rel_err = diff / (abs(m_val) + 1e-30) * 100
        
        status = rel_err < 1e-10 ? "✅" : (rel_err < 1e-6 ? "⚠️ " : "❌")
        @printf("%s M[%d]: MATLAB=%.15e, Julia=%.15e, Diff=%.2e (%.2e%%)\n", 
                status, i, m_val, j_val, diff, rel_err)
    end
    
    # Check for NaN
    if any(isnan.(julia_M))
        println("\n❌ Julia has NaN values!")
    elseif any(isinf.(julia_M))
        println("\n❌ Julia has Inf values!")
    else
        max_err = maximum(abs.(matlab_M .- julia_M) ./ (abs.(matlab_M) .+ 1e-30)) * 100
        if max_err < 1e-10
            println("\n✅ PERFECT MATCH to machine precision!")
        elseif max_err < 1e-6
            @printf("\n✅ GOOD: Max error %.2e%%\n", max_err)
        else
            @printf("\n⚠️  Max error %.2e%%\n", max_err)
        end
    end
end
